python/src/pyunigen.cpp

```diff
@@ -0,0 +1,390 @@
+
+void pybinding_callback(const std::vector<int>& solution, void *source)
+{
+    PyObject** py_obj_cont = (PyObject**) source;
+
+    PyObject* py_list = PyList_New(solution.size());
+    if (py_list == NULL) {
+        PyErr_SetString(PyExc_SystemError, "failed to create a list");
+        return;
+    }
+    for (unsigned int i = 0; i < solution.size(); i++) {
+        PyObject *lit = PyLong_FromLong((long)solution[i]);
+        if (lit == NULL) {
+            PyErr_SetString(PyExc_SystemError, "failed to create a list");
+            return;
+        }
+        PyList_SET_ITEM(py_list, i, lit);
+    }
+
+    (*py_obj_cont) = py_list;
+}
+
+static int parse_sampling_set(Sampler *self, PyObject *sample_set_obj)
+{
+    PyObject *iterator = PyObject_GetIter(sample_set_obj);
+    if (iterator == NULL) {
+        PyErr_SetString(PyExc_TypeError, "iterable object expected");
+        return 1;
+    }
+
+    PyObject *lit;
+    while ((lit = PyIter_Next(iterator)) != NULL) {
+        long val = PyLong_AsLong(lit);
+        if (val == 0) {
+            PyErr_SetString(PyExc_ValueError, "non-zero integer expected");
+            return 1;
+        }
+        if (val > std::numeric_limits<int>::max()/2
+            || val < std::numeric_limits<int>::min()/2
+        ) {
+            PyErr_Format(PyExc_ValueError, "integer %ld is too small or too large", val);
+            return 1;
+        }
+
+        long var = val - 1;
+        self->sampling_set.push_back(var);
+        Py_DECREF(lit);
+    }
+    Py_DECREF(iterator);
+
+    return 0;
+}
+
+static void setup_sampler(Sampler *self, PyObject *args, PyObject *kwds)
+{
+    static char* kwlist[] = {"verbosity", "seed", "kappa", "sampling_set", NULL};
+
+    // All parameters have the same default as the command line defaults
+    // except for verbosity which is 0 by default.
+    self->verbosity = 0;
+    self->seed = 1;
+    self->kappa = 0.638;
+
+    PyObject* sample_set_obj = NULL;
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|iIdO", kwlist,
+        &self->verbosity, &self->seed, &self->kappa, &sample_set_obj))
+    {
+        return;
+    }
+
+    if (sample_set_obj != NULL && parse_sampling_set(self, sample_set_obj)) {
+        return;
+    }
+
+    if (self->verbosity < 0) {
+        PyErr_SetString(PyExc_ValueError, "verbosity must be at least 0");
+        return;
+    }
+    if (self->kappa <= 0 || self->kappa >= 1) {
+        PyErr_SetString(PyExc_ValueError, "epsilon must be greater than 0");
+        return;
+    }
+
+    self->appmc = new ApproxMC::AppMC;
+    self->appmc->set_verbosity(self->verbosity);
+    self->appmc->set_seed(self->seed);
+    self->appmc->set_projection_set(self->sampling_set);
+
+    self->unig = new UniGen::UniG(self->appmc);
+    self->unig->set_verbosity(self->verbosity);
+    self->unig->set_kappa(self->kappa);
+    // Only one sample generated and
+    // multisample disabled for simplicity
+    self->unig->set_multisample(false);
```

Comment:
I would suggest getting multiple samples and so this should be adjusted.

Reply:
Hmm that's an excellent point. Just to clarify, are the multisamples the samples that aren't completely independent, but that you get for "free" from one sample call?

As for the multiple sampling without recounting, the way I set it up you actually call the ApproxMC Python bindings first to get a cell/hash count and then pass the count to Unigen for use in sampling, so once you've counted once you should be able to sample as many times as you want without recounting (I don't think this breaks the independence of the sampling but please let me know if that's not the case!). I did it this way because in the application we're writing that uses ApproxMC/Unigen, we actually want to get the counts once on a bunch of formulas, figure out which ones we want to sample from, and then sample as many times as necessary from all the formulas.

The main reason I had it return one sample instead of multiple is because I'm not super confident using the Python reference counting in C++, so I wanted to avoid nesting too many data structures that might turn into gnarly memory leaks. I can definitely look into adding that to the bindings though, as that would allow for the multisample to work which would definitely be useful for many applications (Assuming I understood how the multisample works above).

Reply:
OK, I see. Let's leave the multisample out for the moment, I'll add that back in later :)